### PR TITLE
Graveler transform catalog interface

### PR DESCRIPTION
**Pull request by nopcoder**

### CODE REVIEW

**Review by guy-har — APPROVED** [review]
Looks good, 
Maybe we should consider returning iterators and not handle pagination in the catalog layer.

File: graveler/graveler.go [hunk 1]
```diff
@@ -131,25 +126,43 @@ type BranchRecord struct {
 	*Branch
 }
 
-// Listing represents either an entry or a CommonPrefix
+// Listing represents CommonPrefix as a key without a Value (nil) or a CommonPrefix as common prefix
```
Comment (johnnyaug) [inline graveler/graveler.go:129]: This is not clear

File: graveler/graveler.go [hunk 2]
```diff
@@ -218,16 +213,21 @@ type Catalog interface {
 	Merge(ctx context.Context, repositoryID RepositoryID, from Ref, to BranchID) (CommitID, error)
 
 	// DiffUncommitted returns the changes as 'Diff' slice on a repository / branch
-	//   List the differences 'from' path, with 'amount' of result.
-	//   Returns differences found, true (boolean) in case there are more differences - use 'from' with last path from previous call to get the next differences
-	DiffUncommitted(ctx context.Context, repositoryID RepositoryID, branchID BranchID, from Path, amount int) ([]Diff, bool, error)
+	//   List the differences 'from' key, with 'amount' of result.
```
Comment (johnnyaug) [inline graveler/graveler.go:216]: amount was removed

File: graveler/graveler.go [hunk 3]
```diff
@@ -218,16 +213,21 @@ type Catalog interface {
...
+	//   Returns differences found, true (boolean) in case there are more differences - use 'from' with last key from previous call to get the next differences
```
Comment (johnnyaug) [inline graveler/graveler.go:217]: bool was removed

File: graveler/graveler.go [hunk 4]
```diff
@@ -218,16 +213,21 @@ type Catalog interface {
...
+	DiffUncommitted(ctx context.Context, repositoryID RepositoryID, branchID BranchID, from Key) (DiffIterator, error)
+
+	// Diff returns the changes between 'left' and 'right' ref, list changes 'from' key with no more than 'amount' per call.
+	//   Returns the list of changes, true (boolean) in case there are more differences - use last key as 'from' in the next call to continue getting differences
+	Diff(ctx context.Context, repositoryID RepositoryID, left, right Ref, from Key) (DiffIterator, error)
```
Comment (johnnyaug) [inline graveler/graveler.go:222]: also here, update docs

File: graveler/graveler.go (hunk 3 above)
Comment (johnnyaug) [reply graveler/graveler.go:217]: do we still want the _from_ in all methods returning iterators? because one can use SeekGE

Comment (nopcoder) [reply graveler/graveler.go:217]: thought about it - when you implement your iterator did you perform a query on NewIterator or waited for the first Next?

Comment (johnnyaug) [reply graveler/graveler.go:217]: waited for next

Comment (nopcoder) [reply graveler/graveler.go:217]: cool - I'll update all

File: graveler/graveler.go [hunk 5]
```diff
@@ -191,9 +196,7 @@ type Catalog interface {
 
 	// ListBranches lists branches on repositories
 	//   The 'from' is used to get all branches after this branch id
```
Comment (johnnyaug) [inline graveler/graveler.go:198]:
```suggestion
```

File: graveler/graveler.go [hunk 6]
```diff
@@ -217,17 +220,20 @@ type Catalog interface {
...
-	// DiffUncommitted returns the changes as 'Diff' slice on a repository / branch
-	//   List the differences 'from' path, with 'amount' of result.
-	//   Returns differences found, true (boolean) in case there are more differences - use 'from' with last path from previous call to get the next differences
-	DiffUncommitted(ctx context.Context, repositoryID RepositoryID, branchID BranchID, from Path, amount int) ([]Diff, bool, error)
+	// DiffUncommitted returns iterator to scan the changes made on the branch
+	DiffUncommitted(ctx context.Context, repositoryID RepositoryID, branchID BranchID) (DiffIterator, error)
+
+	// Diff returns the changes between 'left' and 'right' ref, list changes 'from' key with no more than 'amount' per call.
+	//   Returns the list of changes, true (boolean) in case there are more differences - use last key as 'from' in the next call to continue getting differences
+	Diff(ctx context.Context, repositoryID RepositoryID, left, right Ref, from Key) (DiffIterator, error)
```
Comment (johnnyaug) [inline graveler/graveler.go:228]:
```suggestion
	// Diff returns the changes between 'left' and 'right' ref, starting from the 'from' key .
	Diff(ctx context.Context, repositoryID RepositoryID, left, right Ref, from Key) (DiffIterator, error)
```

**Review by johnnyaug — APPROVED** [review]

**Pull request merged**